neo3/wallet/account.py

```diff
@@ -0,0 +1,283 @@
+from __future__ import annotations
+
+import base58
+import hashlib
+import unicodedata
+from typing import Optional
+
+from Crypto.Cipher import AES
+from neo3 import settings, contracts
+from neo3.core import types, to_script_hash
+from neo3.core.cryptography import ECPoint
+from neo3.core.cryptography import KeyPair
+
+# both constants below are used to encrypt/decrypt a private key to/from a nep2 key
+NEP_HEADER = bytes([0x01, 0x42])
+NEP_FLAG = bytes([0xe0])
+
+
+class Account:
+
+    def __init__(self, password: str,
+                 private_key: Optional[bytes] = None,
+                 watch_only: bool = False,
+                 address: Optional[str] = None
+                 ):
+        """
+        Instantiate an account. This constructor should only be directly called when it's desired to create a new
+        account using just a password and a randomly generated private key, otherwise, they should call another
+        function.
+        """
+
+        public_key: Optional[ECPoint] = None
+        encrypted_key: Optional[bytes] = None
+
+        if watch_only:
+            if address is None:
+                raise ValueError("Creating a watch only account requires an address")
+            elif not self.is_valid_address(address):
+                raise ValueError(f"The given address is not valid. It's size is not {len(types.UInt160.zero()) + 1}"
+                                 f"or the account version is not {settings.network.account_version}")
+
+        if not watch_only:
```

Comment:
1. Can you merge this logic into a single if/else?
```python
if watch_only:
   # address checks
else:
   # key pair stuff
```
2. What is your opinion on the `is_valid_address` function? I wonder if we should add another one `validate_address()` that raises exceptions. 

   The reason being is that I see some almost code duplication,  here (L38)
   ```python
   elif not self.is_valid_address(address):
                   raise ValueError(f"The given address is not valid. It's size is not {len(types.UInt160.zero()) + 1}"
                                    f"or the account version is not {settings.network.account_version}")
   ```

   and this in `address_to_script_hash` (L143)
   ```python
   if not Account.is_valid_address(address):
               if len(data_) != len(types.UInt160.zero()) + 1:
                   raise ValueError(f"The address is wrong, because data_ length should be "
                                    f"{len(types.UInt160.zero()) + 1}")
   
               if data_[0] != settings.network.account_version:
                   raise ValueError(f"The account version is not {settings.network.account_version}")
   ```

   If we add `validate_address()` that throws exceptions, then we can have `is_valid_address` call that internally and catch the exceptions. This is gives one function with a lot of control and one simplified.

Reply:
> If we add validate_address() that throws exceptions, then we can have is_valid_address call that internally and catch the exceptions. This is gives one function with a lot of control and one simplified.

Just to make sure I understood what is being proposed, `is_valid_address()` and `validate_address` should be something like this:
```python
    @staticmethod
    def is_valid_address(address: str) -> bool:
        try:
            Account.validate_address(address)
        except ValueError:
            return False
        return True

    @staticmethod
    def validate_address(address: str):
        data_ = base58.b58decode_check(address)
        if len(data_) != len(types.UInt160.zero()) + 1:
            raise ValueError("")
        elif data_[0] != settings.network.account_version:
            raise ValueError("")
```
And whenever `is_valid_address()` is being called in the code, it should be something like this: 
```python
        if not is_valid_address(address):
            validate_address(address)    # raises the correct exception message
```
If that's the case, then I think that it's ok.

Reply:
Yes exactly!